### PR TITLE
[core] Introduce PredicateUtils to refactor Predicate handle

### PR DESCRIPTION
**Pull request by xuzifu666**
### Purpose

Linked issue: close #xxx
1. Follow up https://github.com/apache/paimon/pull/4436 , refactor related logic to a util class to duplicate codes;
2. Fix incidental ut failed due to order in schema test.

### CODE REVIEW

File: paimon-common/src/main/java/org/apache/paimon/predicate/PredicateUtils.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+     */
+    public static void traverseCompoundPredicate(
+            Predicate predicate,
+            String leafName,
+            Consumer<Predicate> matchConsumer,
+            Consumer<Predicate> unMatchConsumer) {
+        Preconditions.checkState(
+                predicate instanceof CompoundPredicate,
+                "PredicateUtils##handleCompoundPredicate should handle with a CompoundPredicate.");
+
```
Comment (LinMingQiang) [inline paimon-common/src/main/java/org/apache/paimon/predicate/PredicateUtils.java:45]: PredicateUtils.traverseCompoundPredicate only supports processing Predicates of CompoundPredicate type

File: paimon-core/src/main/java/org/apache/paimon/table/system/SchemasTable.java [hunk 2]
```diff
@@ -221,26 +222,28 @@ public InnerTableRead withFilter(Predicate predicate) {
...
-                    List<Predicate> children = compoundPredicate.children();
-                    for (Predicate leaf : children) {
-                        handleLeafPredicate(leaf, leafName);
-                    }
+                    PredicateUtils.traverseCompoundPredicate(
+                            predicate,
+                            leafName,
+                            (Predicate p) -> {
+                                handleLeafPredicate(p, leafName);
+                            },
```
Comment (LinMingQiang) [inline paimon-core/src/main/java/org/apache/paimon/table/system/SchemasTable.java:230]: Can `handleLeafPredicate` method also be extracted to the util?

File: paimon-common/src/main/java/org/apache/paimon/predicate/PredicateUtils.java (hunk 1 above)
Comment (xuzifu666) [reply paimon-common/src/main/java/org/apache/paimon/predicate/PredicateUtils.java:45]: addressed

File: paimon-core/src/main/java/org/apache/paimon/table/system/SchemasTable.java (hunk 2 above)
Comment (xuzifu666) [reply paimon-core/src/main/java/org/apache/paimon/table/system/SchemasTable.java:230]: I had considered about it，this method common logic is less，extract may cost more codes，so keep as current maybe better. Thanks~ @LinMingQiang

**Comment by LinMingQiang** [comment]
+1

File: paimon-core/src/main/java/org/apache/paimon/table/system/SchemasTable.java [hunk 3]
```diff
@@ -221,26 +222,28 @@ public InnerTableRead withFilter(Predicate predicate) {
...
-                    }
+                    PredicateUtils.traverseCompoundPredicate(
+                            predicate,
+                            leafName,
+                            (Predicate p) -> {
+                                if (schemaIds != null) {
+                                    schemaIds.add((Long) ((LeafPredicate) p).literals().get(0));
+                                }
+                            },
+                            (Predicate p) -> {
```
Comment (wwj6591812) [inline paimon-core/src/main/java/org/apache/paimon/table/system/SchemasTable.java:244]: (Predicate p) -> schemaIds = null

File: paimon-core/src/main/java/org/apache/paimon/table/system/SchemasTable.java [hunk 4]
```diff
@@ -221,26 +222,28 @@ public InnerTableRead withFilter(Predicate predicate) {
...
                 CompoundPredicate compoundPredicate = (CompoundPredicate) predicate;
                 if ((compoundPredicate.function()) instanceof And) {
-                    List<Predicate> children = compoundPredicate.children();
-                    for (Predicate leaf : children) {
-                        handleLeafPredicate(leaf, leafName);
-                    }
+                    PredicateUtils.traverseCompoundPredicate(
+                            predicate,
+                            leafName,
+                            (Predicate p) -> {
```
Comment (wwj6591812) [inline paimon-core/src/main/java/org/apache/paimon/table/system/SchemasTable.java:228]: May replaced with expression lambda better?

(Predicate p) -> handleLeafPredicate(p, leafName)

Comment (wwj6591812) [inline paimon-core/src/main/java/org/apache/paimon/table/system/SchemasTable.java:228]: May replaced with expression lambda better?

(Predicate p) -> handleLeafPredicate(p, leafName)

File: paimon-core/src/main/java/org/apache/paimon/table/system/SchemasTable.java (hunk 3 above)
Comment (wwj6591812) [inline paimon-core/src/main/java/org/apache/paimon/table/system/SchemasTable.java:244]: (Predicate p) -> schemaIds = null

File: paimon-core/src/main/java/org/apache/paimon/table/system/SchemasTable.java [hunk 5]
```diff
@@ -221,26 +222,28 @@ public InnerTableRead withFilter(Predicate predicate) {
...
+                                if (schemaIds != null) {
+                                    schemaIds.add((Long) ((LeafPredicate) p).literals().get(0));
+                                }
+                            },
+                            (Predicate p) -> {
+                                schemaIds = null;
```
Comment (wwj6591812) [inline paimon-core/src/main/java/org/apache/paimon/table/system/SchemasTable.java:245]: Why not schemaIds.clear()?

File: paimon-core/src/main/java/org/apache/paimon/table/system/SnapshotsTable.java [hunk 6]
```diff
@@ -223,26 +224,27 @@ public InnerTableRead withFilter(Predicate predicate) {
...
+                    PredicateUtils.traverseCompoundPredicate(
+                            predicate,
+                            leafName,
+                            (Predicate p) -> {
+                                if (snapshotIds != null) {
+                                    snapshotIds.add((Long) ((LeafPredicate) p).literals().get(0));
+                                }
+                            },
+                            (Predicate p) -> {
+                                snapshotIds = null;
```
Comment (wwj6591812) [inline paimon-core/src/main/java/org/apache/paimon/table/system/SnapshotsTable.java:246]: Why not snapshotIds.clear()?

File: paimon-core/src/main/java/org/apache/paimon/table/system/TagsTable.java [hunk 7]
```diff
@@ -239,31 +240,24 @@ public RecordReader<InternalRow> createReader(Split split) {
...
+                                TAG_NAME,
+                                (Predicate p) -> {
+                                    String equalValue =
+                                            ((LeafPredicate) p).literals().get(0).toString();
+                                    if (predicateMap != null && tagManager.tagExists(equalValue)) {
+                                        predicateMap.put(equalValue, tagManager.tag(equalValue));
+                                    }
+                                },
+                                (Predicate p) -> {
+                                    predicateMap = null;
```
Comment (wwj6591812) [inline paimon-core/src/main/java/org/apache/paimon/table/system/TagsTable.java:254]: Why not predicateMap.clear();

File: paimon-core/src/main/java/org/apache/paimon/table/system/SnapshotsTable.java (hunk 6 above)
Comment (xuzifu666) [reply paimon-core/src/main/java/org/apache/paimon/table/system/SnapshotsTable.java:247]: Had add clear method

File: paimon-core/src/main/java/org/apache/paimon/table/system/SchemasTable.java (hunk 3 above)
Comment (xuzifu666) [reply paimon-core/src/main/java/org/apache/paimon/table/system/SchemasTable.java:244]: add clear for it.

**Comment by xuzifu666** [comment]
@wwj6591812 had addressed, Thanks for review.

File: paimon-core/src/main/java/org/apache/paimon/table/system/TagsTable.java (hunk 7 above)
Comment (xuzifu666) [reply paimon-core/src/main/java/org/apache/paimon/table/system/TagsTable.java]: had just set to null which more efficient.

File: paimon-common/src/main/java/org/apache/paimon/predicate/PredicateUtils.java [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+package org.apache.paimon.predicate;
+
+import org.apache.paimon.utils.Preconditions;
+
+import java.util.List;
+import java.util.function.Consumer;
+
+/** A utils to handle {@link Predicate}. */
+public class PredicateUtils {
```
Comment (JingsongLi) [inline paimon-common/src/main/java/org/apache/paimon/predicate/PredicateUtils.java:27]: InPredicateVisitor?

File: paimon-common/src/main/java/org/apache/paimon/predicate/PredicateUtils.java [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+
+    /**
+     * Method for handling with CompoundPredicate.
+     *
+     * @param predicate CompoundPredicate to traverse handle
+     * @param leafName LeafPredicate name
+     * @param matchConsumer leafName matched handle
+     * @param unMatchConsumer leafName unmatched handle
+     */
+    public static void traverseCompoundPredicate(
```
Comment (JingsongLi) [inline paimon-common/src/main/java/org/apache/paimon/predicate/PredicateUtils.java:37]: `extractInElements`

File: paimon-common/src/main/java/org/apache/paimon/predicate/PredicateUtils.java [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+            Predicate predicate,
+            String leafName,
+            Consumer<Predicate> matchConsumer,
+            Consumer<Predicate> unMatchConsumer) {
+        Preconditions.checkArgument(
```
Comment (JingsongLi) [inline paimon-common/src/main/java/org/apache/paimon/predicate/PredicateUtils.java:42]: if (predicate is in predicate): just get elements from it.

File: paimon-common/src/main/java/org/apache/paimon/predicate/PredicateUtils.java (hunk 9 above)
Comment (JingsongLi) [inline paimon-common/src/main/java/org/apache/paimon/predicate/PredicateUtils.java:37]: Just return `Optional<List<Object>>`, elements of in.

**Review by JingsongLi — CHANGES_REQUESTED** [review]

File: paimon-common/src/main/java/org/apache/paimon/predicate/PredicateUtils.java (hunk 9 above)
Comment (xuzifu666) [reply paimon-common/src/main/java/org/apache/paimon/predicate/PredicateUtils.java:37]: Get it, would modify as suggestion，Thanks @JingsongLi

Comment (xuzifu666) [reply paimon-common/src/main/java/org/apache/paimon/predicate/PredicateUtils.java:37]: I'm traveling outside, would modify it in these two days.

Comment (xuzifu666) [reply paimon-common/src/main/java/org/apache/paimon/predicate/PredicateUtils.java:37]: Change the pr to a new pr https://github.com/apache/paimon/pull/4486

**Pull request closed without merging**